### PR TITLE
Refactor to use preexisting AST, support array literals, use AstPrinter

### PR DESCRIPTION
**Pull request by noahlemen**
- Removes forked BinAst versions of AST nodes
- Refactors parser to return normal AST nodes
- Refactors visitors to operate on normal AST nodes
- Refactors test_binast to use AstPrinter instead of requiring our own BinAstPrintVisitor
- removes BinAstPrintVisitor
- removes BinAstValueFactory
- To test changes, adds Array literal test case in test_binast, adds support for this node in visitors
- removes BinAstParseInfo

Still hope to make some of the other parts of this less tedious, but I figured this was enough to try to get out for now.

<details>
<summary>test_binast output</summary>

```
Hello, v8!
v8 initialized

Script #0
'use strict';
42

Creating BinAstParseInfo...Done!
Creating BinAstParser...Done!
Parsing program...Done!
Dumping AST...
FUNC at 0
. KIND 0
. LITERAL ID 0
. SUSPEND COUNT 0
. NAME ""
. INFERRED NAME ""
. EXPRESSION STATEMENT at 0
. . LITERAL "use strict"
. EXPRESSION STATEMENT at 14
. . LITERAL 42


Script #1
'use strict';
'foo'

Creating BinAstParseInfo...Done!
Creating BinAstParser...Done!
Parsing program...Done!
Dumping AST...
FUNC at 0
. KIND 0
. LITERAL ID 0
. SUSPEND COUNT 0
. NAME ""
. INFERRED NAME ""
. EXPRESSION STATEMENT at 0
. . LITERAL "use strict"
. EXPRESSION STATEMENT at 14
. . LITERAL "foo"


Script #2
'use strict';
if (true) { 'foo'; }

Creating BinAstParseInfo...Done!
Creating BinAstParser...Done!
Parsing program...Done!
Dumping AST...
FUNC at 0
. KIND 0
. LITERAL ID 0
. SUSPEND COUNT 0
. NAME ""
. INFERRED NAME ""
. EXPRESSION STATEMENT at 0
. . LITERAL "use strict"
. IF at 14
. . CONDITION at 18
. . . LITERAL true
. . THEN at -1
. . . BLOCK at -1
. . . . EXPRESSION STATEMENT at 26
. . . . . LITERAL "foo"


Script #3
'use strict';
if (true) { 'foo'; } else { 'bar'; }

Creating BinAstParseInfo...Done!
Creating BinAstParser...Done!
Parsing program...Done!
Dumping AST...
FUNC at 0
. KIND 0
. LITERAL ID 0
. SUSPEND COUNT 0
. NAME ""
. INFERRED NAME ""
. EXPRESSION STATEMENT at 0
. . LITERAL "use strict"
. IF at 14
. . CONDITION at 18
. . . LITERAL true
. . THEN at -1
. . . BLOCK at -1
. . . . EXPRESSION STATEMENT at 26
. . . . . LITERAL "foo"
. . ELSE at -1
. . . BLOCK at -1
. . . . EXPRESSION STATEMENT at 42
. . . . . LITERAL "bar"


Script #4
'use strict';
var sum = 0;
for (var i = 0; i < 10; ++i) {
  sum += i;
}
console.log(sum);

Creating BinAstParseInfo...Done!
Creating BinAstParser...Done!
Parsing program...Done!
Dumping AST...
FUNC at 0
. KIND 0
. LITERAL ID 0
. SUSPEND COUNT 0
. NAME ""
. INFERRED NAME ""
. DECLS
. . VARIABLE (0x10686dcf0) (mode = VAR, assigned = true) "sum"
. . VARIABLE (0x10686ddf8) (mode = VAR, assigned = true) "i"
. EXPRESSION STATEMENT at 0
. . LITERAL "use strict"
. BLOCK NOCOMPLETIONS at -1
. . EXPRESSION STATEMENT at 24
. . . INIT at 24
. . . . VAR PROXY unallocated (0x10686dcf0) (mode = VAR, assigned = true) "sum"
. . . . LITERAL 0
. FOR at 27
. . INIT at -1
. . . BLOCK NOCOMPLETIONS at -1
. . . . EXPRESSION STATEMENT at 40
. . . . . INIT at 40
. . . . . . VAR PROXY unallocated (0x10686ddf8) (mode = VAR, assigned = true) "i"
. . . . . . LITERAL 0
. . COND at 45
. . . LT at 45
. . . . VAR PROXY unresolved "i"
. . . . LITERAL 10
. . BODY at -1
. . . BLOCK at -1
. . . . EXPRESSION STATEMENT at 60
. . . . . ASSIGN_ADD at 64
. . . . . . VAR PROXY unresolved "sum"
. . . . . . VAR PROXY unresolved "i"
. . NEXT at 53
. . . EXPRESSION STATEMENT at 53
. . . . PRE INC at 53
. . . . . VAR PROXY unresolved "i"
. EXPRESSION STATEMENT at 72
. . CALL
. . . PROPERTY at 80
. . . . VAR PROXY unresolved "console"
. . . . NAME log
. . . VAR PROXY unresolved "sum"


Script #5
'use strict';
var square = function(x) { return x * x; };
console.log('square(3) =', square(3));

Creating BinAstParseInfo...Done!
Creating BinAstParser...Done!
Parsing program...Done!
Dumping AST...
FUNC at 0
. KIND 0
. LITERAL ID 0
. SUSPEND COUNT 0
. NAME ""
. INFERRED NAME ""
. DECLS
. . VARIABLE (0x12401aaf0) (mode = VAR, assigned = true) "square"
. EXPRESSION STATEMENT at 0
. . LITERAL "use strict"
. BLOCK NOCOMPLETIONS at -1
. . EXPRESSION STATEMENT at 27
. . . INIT at 27
. . . . VAR PROXY unallocated (0x12401aaf0) (mode = VAR, assigned = true) "square"
. . . . FUNC LITERAL at 27
. . . . . LITERAL ID 1
. . . . . NAME square
. . . . . INFERRED NAME square
. EXPRESSION STATEMENT at 58
. . CALL
. . . PROPERTY at 66
. . . . VAR PROXY unresolved "console"
. . . . NAME log
. . . LITERAL "square(3) ="
. . . CALL
. . . . VAR PROXY unresolved "square"
. . . . LITERAL 3


Script #6
'use strict';
function square(x) { return x * x; }
console.log('square(5) =', square(5));

Creating BinAstParseInfo...Done!
Creating BinAstParser...Done!
Parsing program...Done!
Dumping AST...
FUNC at 0
. KIND 0
. LITERAL ID 0
. SUSPEND COUNT 0
. NAME ""
. INFERRED NAME ""
. DECLS
. . FUNCTION "square" = function square
. EXPRESSION STATEMENT at 0
. . LITERAL "use strict"
. EXPRESSION STATEMENT at 51
. . CALL
. . . PROPERTY at 59
. . . . VAR PROXY unresolved "console"
. . . . NAME log
. . . LITERAL "square(5) ="
. . . CALL
. . . . VAR PROXY unresolved "square"
. . . . LITERAL 5


Script #7
'use strict';
var obj = {};

Creating BinAstParseInfo...Done!
Creating BinAstParser...Done!
Parsing program...Done!
Dumping AST...
FUNC at 0
. KIND 0
. LITERAL ID 0
. SUSPEND COUNT 0
. NAME ""
. INFERRED NAME ""
. DECLS
. . VARIABLE (0x12401aaf0) (mode = VAR, assigned = true) "obj"
. EXPRESSION STATEMENT at 0
. . LITERAL "use strict"
. BLOCK NOCOMPLETIONS at -1
. . EXPRESSION STATEMENT at 24
. . . INIT at 24
. . . . VAR PROXY unallocated (0x12401aaf0) (mode = VAR, assigned = true) "obj"
. . . . OBJ LITERAL at 24


Script #8
'use strict';
var obj = {foo: 'bar', bar: 1, baz: function(x) { return x; }};

Creating BinAstParseInfo...Done!
Creating BinAstParser...Done!
Parsing program...Done!
Dumping AST...
FUNC at 0
. KIND 0
. LITERAL ID 0
. SUSPEND COUNT 0
. NAME ""
. INFERRED NAME ""
. DECLS
. . VARIABLE (0x12401aaf0) (mode = VAR, assigned = true) "obj"
. EXPRESSION STATEMENT at 0
. . LITERAL "use strict"
. BLOCK NOCOMPLETIONS at -1
. . EXPRESSION STATEMENT at 24
. . . INIT at 24
. . . . VAR PROXY unallocated (0x12401aaf0) (mode = VAR, assigned = true) "obj"
. . . . OBJ LITERAL at 24
. . . . . PROPERTY - CONSTANT
. . . . . . KEY at 25
. . . . . . . LITERAL "foo"
. . . . . . VALUE at 30
. . . . . . . LITERAL "bar"
. . . . . PROPERTY - CONSTANT
. . . . . . KEY at 37
. . . . . . . LITERAL "bar"
. . . . . . VALUE at 42
. . . . . . . LITERAL 1
. . . . . PROPERTY - COMPUTED
. . . . . . KEY at 45
. . . . . . . LITERAL "baz"
. . . . . . VALUE at 50
. . . . . . . FUNC LITERAL at 50
. . . . . . . . LITERAL ID 1
. . . . . . . . NAME baz
. . . . . . . . INFERRED NAME obj.baz


Script #9
'use strict';
var obj = {}; obj.foo = 'bar';

Creating BinAstParseInfo...Done!
Creating BinAstParser...Done!
Parsing program...Done!
Dumping AST...
FUNC at 0
. KIND 0
. LITERAL ID 0
. SUSPEND COUNT 0
. NAME ""
. INFERRED NAME ""
. DECLS
. . VARIABLE (0x12401aaf0) (mode = VAR, assigned = true) "obj"
. EXPRESSION STATEMENT at 0
. . LITERAL "use strict"
. BLOCK NOCOMPLETIONS at -1
. . EXPRESSION STATEMENT at 24
. . . INIT at 24
. . . . VAR PROXY unallocated (0x12401aaf0) (mode = VAR, assigned = true) "obj"
. . . . OBJ LITERAL at 24
. EXPRESSION STATEMENT at 28
. . ASSIGN at 36
. . . PROPERTY at 32
. . . . VAR PROXY unresolved "obj"
. . . . NAME foo
. . . LITERAL "bar"


Script #10
'use strict';
var arr = [];

Creating BinAstParseInfo...Done!
Creating BinAstParser...Done!
Parsing program...Done!
Dumping AST...
FUNC at 0
. KIND 0
. LITERAL ID 0
. SUSPEND COUNT 0
. NAME ""
. INFERRED NAME ""
. DECLS
. . VARIABLE (0x12401aaf0) (mode = VAR, assigned = true) "arr"
. EXPRESSION STATEMENT at 0
. . LITERAL "use strict"
. BLOCK NOCOMPLETIONS at -1
. . EXPRESSION STATEMENT at 24
. . . INIT at 24
. . . . VAR PROXY unallocated (0x12401aaf0) (mode = VAR, assigned = true) "arr"
. . . . ARRAY LITERAL at 24

Process 59492 exited with status = 0 (0x00000000)
```
</details>

<details>
<summary>d8 test.js output</summary>

```
Deserialized function literal in 34 us: 0x1038092d0
Deserialized function literal in 34 us: 0x10380c448
Deserialized function literal in 26 us: 0x1043f8560
running callback
Deserialized function literal in 60 us: 0x1043f8470
84
Deserialized function literal in 79 us: 0x1043f84b0
72
Process 60122 exited with status = 0 (0x00000000)
```
</details>

### CODE REVIEW

File: src/ast/ast.h
Comment (noahlemen) [inline src/ast/ast.h:158]: needed for detecting unimplemented nodes in visitor

File: src/parsing/binast-serialize-visitor.h
Comment (noahlemen) [inline src/parsing/binast-serialize-visitor.h]: was this incorrect? since insert is called on this in loops in SerializeStringTable, it seems guaranteed to be nonzero at some point. was encountering this issue when running test_binast

File: BUILD.gn
Comment (noahlemen) [inline BUILD.gn]: lol whoops

File: src/parsing/binast-serialize-visitor.h
Comment (noahlemen) [inline src/parsing/binast-serialize-visitor.h:187]: (i could just leave this commented out or otherwise not submit this change if this is still helpful for the output for you)

Comment (mhahnenberg) [reply src/parsing/binast-serialize-visitor.h:187]: Nah, that's okay. If I need it again I can re-add it.

File: src/parsing/binast-serialize-visitor.h
Comment (mhahnenberg) [reply src/parsing/binast-serialize-visitor.h]: It's testing for the presence of the string in the table. We should only ever serialize an AstRawString once. This assert firing means that there's something wrong with that assumption, so yeah you'll need to debug why this is happening b/c it shouldn't be.

**Review by mhahnenberg — CHANGES_REQUESTED** [review]
Everything looks good except for that assertion firing--that invariant should hold.

File: src/parsing/binast-serialize-visitor.h
Comment (mhahnenberg) [reply src/parsing/binast-serialize-visitor.h]: My first guess would be that we're reusing the same BinAstSerializeVisitor in multiple contexts, i.e. `string_table_indices_` was already filled in by something and then we tried to add the same basic constant strings again.

Comment (mhahnenberg) [reply src/parsing/binast-serialize-visitor.h]: Turns out my guess was wrong :sweat_smile:

**Review by mhahnenberg — APPROVED** [review]
Cool, everything looks good!

**Comment by noahlemen** [comment]
d8 test seems to be failing on this diff after rebase (but not on base), need to investigate before landing

**Review by mhahnenberg — APPROVED** [review]
LGTM!

**Pull request merged**
